### PR TITLE
Skip shapes which don't block any edges

### PR DESCRIPTION
**Pull request by otbutz**
BlockArea currently treats an empty GHIntHashSet the same way as a non-existing cache for blocked edges. But a small shape which doesn't cover edges in the graph would also yield an empty GHIntHashSet. For those cases BlockArea would resort to the more expensive bbox/shape intersection logic:

https://github.com/graphhopper/graphhopper/blob/0522c495ccad5f22fe19c6e858a40a7bd6f74eae/core/src/main/java/com/graphhopper/storage/GraphEdgeIdFinder.java#L198-L223

I changed the logic to explicitly store `null` as the cache for blocked edges if the shape is to large so we can properly determine if we have no cache or if it's just empty.

### CODE REVIEW

**Comment by karussell** [comment]
If a shape does not cover edges - maybe we should even throw an exception? (it might be too ugly?)

Did you had a chance to measure the query difference? Because this API is kind of legacy and if the difference is huge it will be critical to implement this for the new option via custom model too.

**Comment by otbutz** [comment]
Screenshot_20210506_111718

**Comment by otbutz** [comment]
Noticed this while benchmarking requests with a few blocked points. And e.g a point might not always cover an edge in the graph if the road is not present in OSM.

**Comment by otbutz** [comment]
We could also drop these kind of shapes. There's no point to do the checks if we know that we won't find any intersections?

**Comment by karussell** [comment]
Yes, either this or throwing an exception.

**Comment by otbutz** [comment]
Apart from `GraphHopperTest:556` everything seems to work.

**Comment by otbutz** [comment]
Shouldn't have skipped my :coffee: Readded the `continue` statement in the intersect check loop and everything is working as expected

File: core/src/main/java/com/graphhopper/storage/GraphEdgeIdFinder.java [hunk 1]
```diff
@@ -165,23 +175,26 @@ public BlockArea(Graph g) {
...
 
         public String toString(int shapeIndex) {
+            GHIntHashSet blockedEdges = edgesList.get(shapeIndex);
+            if (blockedEdges == null) {
+                return Collections.emptyList().toString();
+            }
+            
             List<Integer> returnList = new ArrayList<>();
-            for (IntCursor intCursor : edgesList.get(shapeIndex)) {
+            for (IntCursor intCursor : blockedEdges) {
```
Comment (karussell) [inline core/src/main/java/com/graphhopper/storage/GraphEdgeIdFinder.java:188]: Are these changes necessary anymore? Because the edgesList contains now only HashSets that are != null and with size > 0 after your recent changes. (?)

Comment (otbutz) [reply core/src/main/java/com/graphhopper/storage/GraphEdgeIdFinder.java:188]: Yes. We either store `null` HashSets for large Shapes or non-empty ones for small shapes.

Comment (otbutz) [reply core/src/main/java/com/graphhopper/storage/GraphEdgeIdFinder.java:188]: Yes. We either store a `null` HashSet for large shapes or non-empty HashSets for smaller ones. We only drop empty HashSets for small shapes.

File: core/src/main/java/com/graphhopper/storage/GraphEdgeIdFinder.java [hunk 2]
```diff
@@ -200,8 +213,8 @@ public final boolean intersects(EdgeIteratorState edgeState) {
             BBox bbox = null;
             for (int shapeIdx = 0; shapeIdx < blockedShapes.size(); shapeIdx++) {
                 GHIntHashSet blockedEdges = edgesList.get(shapeIdx);
-                // blockedEdges acts as cache that is only useful when filled and for non-virtual edges
-                if (!blockedEdges.isEmpty() && edgeState.getEdge() < baseEdgeCount) {
+                // blockedEdges acts as cache if present and is only useful for non-virtual edges
+                if (blockedEdges != null && edgeState.getEdge() < baseEdgeCount) {
```
[the remark this answers is not in the document]
Comment (otbutz) [reply core/src/main/java/com/graphhopper/storage/GraphEdgeIdFinder.java:217]: Throwing an exception for empty Sets sounds good

File: core/src/main/java/com/graphhopper/storage/GraphEdgeIdFinder.java (hunk 1 above)
Comment (karussell) [reply core/src/main/java/com/graphhopper/storage/GraphEdgeIdFinder.java:188]: > We either store a null HashSet for large shapes

Why not an empty set where we can avoid the special `null` handling?

File: core/src/main/java/com/graphhopper/storage/GraphEdgeIdFinder.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (karussell) [reply core/src/main/java/com/graphhopper/storage/GraphEdgeIdFinder.java:220]: Hm. I just thought about this. The problem then is that users might rely on the silence here. E.g. if users draw arbitrary areas to exclude and before accepting a new blocked area they would need to query us via a route, which sounds ugly. Also it might happen that the area does work now but fails in the future because road is under construction or certain roads are excluded by default and then all routing requests with the empty block area will fail.

**Review by karussell — APPROVED** [review]
Thank you - LGTM!

**Comment by otbutz** [comment]
@karussell the id lookup currently only works for non-virtual edges. Could we use the same mechanism as in `EdgeIdDetails` ?

```java
int edge;
if (edgeState instanceof VirtualEdgeIteratorState)
   edge = GHUtility.getEdgeFromEdgeKey(((VirtualEdgeIteratorState) edgeState).originalEdgeKey());
else
   edge = edgeState.getEdge();
```
